src/main/common/utils.js

```diff
@@ -219,6 +220,72 @@ gpii.app.findButtonById = function (buttonId, availableButtons) {
     return false;
 };
 
+/**
+ * Generates the proper service button schema for the custom button
+ * @param {Object} buttonData - a simple data object with the values needed for the custom button
+ * @return {Object|Boolean} - data object constructed exactly as every other in the settings.json or false
+ */
+gpii.app.generateCustomButton = function (buttonData) {
+    var serviceButtonTypeApp = "custom-launch-app",
+        serviceButtonTypeWeb = "custom-open-url",
+        titleAppNotFound = "Program not found",
+        titleUrlInvalid = "Invalid URL",
+        disabledStyle = "disabledButton",
+        data = false;
+
+    // we need to have the data, with all required fields:
+    // buttonId, buttonName, buttonType, buttonData
+    if (buttonData && fluid.isValue(buttonData.buttonId) && fluid.isValue(buttonData.buttonName) && fluid.isValue(buttonData.buttonType) && fluid.isValue(buttonData.buttonData)) {
+        var buttonType = buttonData.buttonType === "APP" ? serviceButtonTypeApp : serviceButtonTypeWeb;
```

Comment:
We should supply some helpful diagnostic if not all the data is present. You might consider rolling a little utility like https://github.com/fluid-project/infusion/blob/master/src/framework/core/js/Fluid.js#L226